### PR TITLE
Call Python3 linter on custom checks

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

In preparation to python2 being EOL we now ship a Python3 linter in preparation for Python3.

### CODE REVIEW

File: pkg/collector/py/py3_checker.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+type warning struct {
+	Message string
+}
+
+// verifyPython3 checks that a check can run on python 3
+func validatePython3(moduleName string, modulePath string) ([]string, error) {
+	if py3LinterPath == "" {
+		return nil, nil
```
Comment (masci) [inline pkg/collector/py/py3_checker.go]: I think we should return an error here, if some code called `validatePython3` and something went wrong at `init` time, the caller should check whether `res` is `nil` to see if the validation happened or not, which seems a bit weird.

File: pkg/collector/py/loader.go [hunk 2]
```diff
@@ -215,3 +248,21 @@ func addExpvarConfigureError(check string, errMsg string) {
...
+}
+
+func addExpvarPy3Warnings(check string, warnings []string) {
+	statsLock.Lock()
+	defer statsLock.Unlock()
+
+	if len(warnings) > 0 {
+		py3Warnings[check] = warnings
+	} else {
+		log.Debugf("check '%s' seems to be compatible with python3", check)
```
Comment (masci) [inline pkg/collector/py/loader.go]: nit: `Python3`

File: releasenotes/notes/add-py3-linter-bc627c4f509d90f9.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    In preparation to python2 being EOL we now ship a Python3 linter to warn
```
Comment (masci) [inline releasenotes/notes/add-py3-linter-bc627c4f509d90f9.yaml]: nit: `Python2`

**Comment by hush-hush** [comment]
@masci 
> So instead of "Python3 Incompatible Checks" something on the line of "Agent 7.0 incompatible checks"?

I agree with your point. Do you think we should add sections in `Agent 7.0 incompatible checks`, like `python syntax` or something like that. My point is that it could be a bit unintuitive for users to find this the status page:

```
  Agent 7.0 incompatible checks
  ==============================

      custom_check
      ------------
        - Line 7, column 8: print statement used
```
We might want to explain somewhere that it's related to moving to Python3.

What do you think about something like this:
```
  Agent 7.0 incompatible checks
  ===========================

      Python3 issues
      ==============

          custom_check
          ------------
            - Line 7, column 8: print statement used
```

File: pkg/collector/py/py3_checker_nix.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+package py
+
+import (
+	"path/filepath"
+)
+
+const (
+	// Waiting for final name
+	py3LinterBin = "a7_validate"
```
Comment (derekwbrown) [inline pkg/collector/py/py3_checker_nix.go]: Do we know the name yet?

**Review by derekwbrown — DISMISSED** [review]
Looks good to me; is the intent to fix the final `py3LinterBin` in a subsequent PR?

File: pkg/collector/py/py3_checker_nix.go (hunk 4 above)
Comment (hush-hush) [reply pkg/collector/py/py3_checker_nix.go]: yes, that's the final name. Forgot to remove the comment !

**Review by remicalixte — COMMENTED** [review]
LGTM, but I agree the warnings aren't very clear without mentioning python3 explicitly.

**Comment by hush-hush** [comment]
@remicalixte Yes, we're waiting for the exact wording from Product !

File: releasenotes/notes/add-py3-linter-bc627c4f509d90f9.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+---
+features:
+  - |
+    In preparation to Python2 being EOL we now ship a Python3 linter to warn
+    user if they are using custom checks not compatible wiht both version 2 and
```
Comment (gzussa) [inline releasenotes/notes/add-py3-linter-bc627c4f509d90f9.yaml]: typo at `wiht`

**Review by gzussa — COMMENTED** [review]
Great PR!

**Review by olivielpeau — COMMENTED** [review]
Left a few comments, overall LGTM

A couple of additional notes:
* not sure if sending 1/0 is the best option, sending `1` with a `status:` (or `ready:`) tag could make things easier to query
* does this have any measurable performance impact?

File: pkg/collector/py/loader.go [hunk 6]
```diff
@@ -160,7 +185,39 @@ func (cl *PythonCheckLoader) Load(config integration.Config) ([]check.Check, err
 			} else {
 				log.Debugf("python check '%s' doesn't have a '__version__' attribute: %s", config.Name, errors.New(pyErr))
 			}
-			log.Infof("python check '%s' doesn't have a '__version__' attribute", config.Name)
+			log.Debugf("python check '%s' doesn't have a '__version__' attribute", config.Name)
+
+		}
+
+		if !agentConfig.Datadog.GetBool("disable_py3_validation") && !loadedAsWheel {
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply pkg/collector/py/loader.go]: I think having the option of disabling it entirely is always good to have.

File: pkg/collector/py/py3_checker.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+type warning struct {
+	Message string
+}
+
+// verifyPython3 checks that a check can run on python 3
+func validatePython3(moduleName string, modulePath string) ([]string, error) {
+	if py3LinterPath == "" {
+		return nil, fmt.Errorf("no Python3 linter found")
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply pkg/collector/py/py3_checker.go]: This should never happend, so it's useful to log it and send the telemetry I think

File: omnibus/config/software/datadog-pip-agent7.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+name 'datadog-pip-agent7'
+
+dependency 'datadog-pip'
+dependency 'datadog-agent'
+
+version="0.0.3"
+
+build do
+  pip "install datadog-a7==#{version}"
```
Comment (olivielpeau) [inline omnibus/config/software/datadog-pip-agent7.rb]: any reason we're not using the software definition in `omnibus-software`?

File: omnibus/config/software/datadog-pip-agent7.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+# under the Apache License Version 2.0.
+# This product includes software developed at Datadog (https:#www.datadoghq.com/).
+# Copyright 2018 Datadog, Inc.
+
+name 'datadog-pip-agent7'
+
+dependency 'datadog-pip'
+dependency 'datadog-agent'
+
+version="0.0.3"
```
Comment (olivielpeau) [inline omnibus/config/software/datadog-pip-agent7.rb]: should be `0.0.4` now

File: pkg/collector/py/loader.go [hunk 10]
```diff
@@ -160,7 +188,45 @@ func (cl *PythonCheckLoader) Load(config integration.Config) ([]check.Check, err
...
+						tags = append(tags, fmt.Sprintf("status:%s", status))
+						cl.telemetry.Gauge(a7ReadyMetric, val, "", tags)
+					} else {
+						tags = append(tags, fmt.Sprintf("status:%s", a7TagUnknown))
+						cl.telemetry.Gauge(a7ReadyMetric, a7NotReady, "", tags)
+						log.Warnf("Could not lint check %s for Python3 compatibility: %s", name, err)
+					}
+				} else {
+					tags = append(tags, fmt.Sprintf("status:%s", a7TagUnknown))
+					cl.telemetry.Gauge(a7ReadyMetric, a7NotReady, "", tags)
```
Comment (olivielpeau) [inline pkg/collector/py/loader.go]: I think we should not send the value `a7NotReady` when the status is unknown, even if we do set an "unknown" tag. I think it'd be too easy to confuse it with the "not ready" case.

We may want to send a different value altogether when the validation itself failed to complete though

**Review by olivielpeau — COMMENTED** [review]
left another comment. Let's also remove all the user-facing warnings for now.

File: pkg/collector/py/loader.go (hunk 10 above)
Comment (hush-hush) [reply pkg/collector/py/loader.go]: I added the `unknown` tag. I think it's safe to say that if the linter failed on that check it's not going to run with python3. Wdyt ?

Comment (masci) [reply pkg/collector/py/loader.go]: I think if the linter fails we should assume the worst and blame the python module, +1 for Maxime's solution

Comment (olivielpeau) [reply pkg/collector/py/loader.go]: ok, works for me 👍

File: pkg/status/dist/templates/collector.tmpl [hunk 11]
```diff
@@ -53,7 +53,17 @@ Collector
       {{- end }}
     {{- end}}
   {{- end }}
-{{- end }}
+
+  {{- if .Py3Warnings }}
+  Python 3 Incompatibility
```
Comment (olivielpeau) [inline pkg/status/dist/templates/collector.tmpl]: we should not display anything on the CLI and GUI status page for now

Comment (truthbk) [reply pkg/status/dist/templates/collector.tmpl]: +1

**Review by truthbk — COMMENTED** [review]
Added a small suggestion regarding the way we run the linter. Looks pretty much ready.

There's some stuff here we'll use in the future, but not currently - should we clean up? Definitely fine if it's left in for the RC, pending some discussion. I'm generally not a fan of leaving lingering code, but I see the point here.

File: omnibus/config/software/datadog-pip-agent7.rb (hunk 8 above)
Comment (hush-hush) [reply omnibus/config/software/datadog-pip-agent7.rb]: Fixed

**Review by truthbk — APPROVED** [review]
Looks good to me! ✅

**Pull request merged**